### PR TITLE
Bulk column transformations and data cleansing gems

### PR DESCRIPTION
**Pull request by kathweinschenkprophecy**
To do: Integrate version requirement info when Ashish gets back to me. I think we can merge this first, and add that later.

- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208789833004301
  - https://app.asana.com/0/0/1208789699002283

### CODE REVIEW

File: docs/Spark/gems/transform/bulk-column-expressions.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+---
+
+The Bulk Column Expressions Gem primarily lets you change the data type of multiple columns at once. It provides additional functionality as well, such as renaming... 
+
+
+# Parameters
+
+| Parameter                                    | Description                                                             | Required |
+|----------------------------------------------|-------------------------------------------------------------------------|----------|
+| Data Type of the columns to do operations on | The data type of columns to select.                                     | False    |
```
Comment (kathweinschenkprophecy) [inline docs/Spark/gems/transform/bulk-column-expressions.md:19]:
```suggestion
| Data Type of the columns to do operations on | The data type of columns to select.                                     | True     |
```

File: docs/Spark/gems/transform/bulk-column-expressions.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+| Selected Columns                             | The columns on which to apply transformations.                          | False    |
+| Change output column name                    | An option to add a prefix or suffix to the selected column names.       | False    |
+| Copy incoming columns to output              | When "Change output column name" is selected, this is the option to ... | False    |
+| Change output column type                    | The data type that the columns will be transformed into.                | False    |
+| Output Expression                            | A Spark SQL expression that can be applied to the selected columns.     | False    |
```
Comment (kathweinschenkprophecy) [inline docs/Spark/gems/transform/bulk-column-expressions.md:24]: Example?

File: docs/Spark/gems/transform/data-cleansing.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  - clean
+  - format
+---
+
+Use the Data Cleansing Gem to standardize data formats and address missing or null values in the data.
+
+## Parameters
+| Parameter                        | Description                                                                                              | Required |
+|----------------------------------|----------------------------------------------------------------------------------------------------------|----------|
+| Select columns you want to clean | The set of columns on which to perform cleaning transformations. This is also the output set of columns. | False    |
```
Comment (kathweinschenkprophecy) [inline docs/Spark/gems/transform/data-cleansing.md:17]: Is this supposed to limit the output columns, too?

File: docs/Spark/gems/transform/data-cleansing.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+| Remove null data                 | The method used to remove null data.                                                                     | True     |
```
Comment (kathweinschenkprophecy) [inline docs/Spark/gems/transform/data-cleansing.md:18]: Technically this is not required (doesn't throw an error), but it is automatically populated.

File: docs/Spark/gems/transform/dynamic-select.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+### Remove date columns using field type
+
+Assume you would like to remove irrelevant date and timestamp columns from your dataset. You can do so by selecting all field types to maintain, except for date and timestamp.
+
+<!-- ![Keep all columns except Date and Timestamp column using the visual interface](./img/remove-date-timestamp.png) -->
+
+### Remove date columns with an expression
+
+Using the same example, you can also accomplish this using the expression `column_type != date`.
```
Comment (kathweinschenkprophecy) [inline docs/Spark/gems/transform/dynamic-select.md:35]: how to format this expression?

File: docs/Spark/gems/transform/bulk-column-rename.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+One example may be to add the prefix `meta_` to tag columns that contain metadata.
+
+<!-- ![Add prefix to multiple columns](./img/bulk-add-prefix.png) -->
+
+
+### Use a custom expression
+
+You can accomplish the same or more complex changes using a custom expression like `concat('meta_', column_name)`.
+
+<!-- ![Add prefix to multiple columns using a custom expression](./img/bulk-add-prefix.png) -->
```
Comment (kathweinschenkprophecy) [inline docs/Spark/gems/transform/bulk-column-rename.md:36]: what's the difference between this custom expression and other expression field (dynamic select and bulk expression)? are there limitations between them?

File: docs/Spark/gems/transform/dynamic-select.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+### Remove date columns using field type
+
+Assume you would like to remove irrelevant date and timestamp columns from your Dataset. You can do so with the **Select field types** method by selecting all field types to maintain, except for date and timestamp.
+
+![Keep all columns except Date and Timestamp column using the visual interface](./img/remove-date-timestamp.png)
+
+### Remove date columns with an expression
+
+Using the same example, you can accomplish the same task with the **Select via expression** method by inputting the the expression `column_type NOT IN ('date', 'timestamp')`.
```
Comment (alexanderahn) [inline docs/Spark/gems/transform/dynamic-select.md:35]: Is the Select via expression the same kind of method used in the Bulk Column Rename example?

**Review by alexanderahn — APPROVED** [review]
LGTM

**Comment by alexanderahn** [comment]
Looks good, I just had a question about the example you used.

Also, do you think stylistically if we should avoid using periods at the end of the Description column lines and unordered list items? https://developers.google.com/style/lists#capitalization-and-end-punctuation

**Comment by alexanderahn** [comment]
> Looks good, I just had a question about the example you used.
> 
> Also, do you think stylistically if we should avoid using periods at the end of the Description column lines and unordered list items? https://developers.google.com/style/lists#capitalization-and-end-punctuation

Actually, the unordered list would be fine in this case, but maybe not the Description lines. WDYT?

File: docs/Spark/gems/transform/dynamic-select.md (hunk 7 above)
Comment (kathweinschenkprophecy) [reply docs/Spark/gems/transform/dynamic-select.md:35]: Good question! I asked the same to Ashish. They are not the same. For both of them, you only have access to a limited number of variables (which actually are included in the UI for dynamic select). Ashish asked me to open an Asana ticket to ask him to include the available variables for bulk column rename.
image

**Comment by kathweinschenkprophecy** [comment]
> Looks good, I just had a question about the example you used.
> 
> Also, do you think stylistically if we should avoid using periods at the end of the Description column lines and unordered list items? https://developers.google.com/style/lists#capitalization-and-end-punctuation

Good point! Maybe we can adopt the same logic as the list style that you linked. In other words, use periods in tables if the sentence includes a verb, but otherwise don't use a period.

**Review by alexanderahn — APPROVED** [review]

**Pull request merged**